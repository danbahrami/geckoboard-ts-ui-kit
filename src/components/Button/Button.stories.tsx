import { Story, Meta } from '@storybook/react';

import { Button, ButtonProps } from './Button';

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: {},
} as Meta<typeof Button>;

const Template: Story<ButtonProps> = (args) => (
  <Button {...args}>Click me</Button>
);

export const Primary = Template.bind({});
Primary.args = {};

export const Secondary = Template.bind({});
Secondary.args = {
  intent: 'secondary',
};

export const Negative = Template.bind({});
Negative.args = {
  intent: 'negative',
};

export const ExtraSmall = Template.bind({});
ExtraSmall.args = {
  size: 'extra-small',
};

export const Small = Template.bind({});
Small.args = {
  size: 'small',
};

export const Large = Template.bind({});
Large.args = {
  size: 'large',
};
