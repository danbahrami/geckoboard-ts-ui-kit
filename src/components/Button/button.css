.Button {
  appearance: none;
  border: none;
  background: green;
  border-radius: 6px;
  font-size: 1rem;
  color: white;
  padding: 0.5em 1em;
  position: relative;
  margin: 0;
  cursor: pointer;
  text-decoration: none !important;

  box-sizing: border-box;
}

.Button:disabled {
  opacity: 0.5;
  cursor: default;
}

.Button:focus {
  box-shadow: 0 0 0 1px #16113a;
}

.Button::-moz-focus-inner {
  border: 0;
  padding: 0;
  margin: 0;
}

/**
 * INTENTS
 **/

/* Primary */
.Button--intent-primary {
  background-color: #0fd354;
  color: #16113a;
}

.Button--intent-primary:visited,
.Button--intent-primary:hover:not(:disabled),
.Button--intent-primary:link,
.Button--intent-primary:focus {
  color: #16113a;
}

.Button--intent-primary:hover:not(:disabled) {
  /* Prevent global button and anchor styles affecting the color */
  background-color: #0ca341;
  color: #16113a;
}

/* Secondary */
.Button--intent-secondary {
  background-color: #dfe4ee;
  color: #16113a;
}

.Button--intent-secondary:visited,
.Button--intent-secondary:hover:not(:disabled),
.Button--intent-secondary:link,
.Button--intent-secondary:focus {
  color: #16113a;
}

.Button--intent-secondary:hover:not(:disabled) {
  /* Prevent global button and anchor styles affecting the color */
  background-color: #bec8dc;
  color: #16113a;
}

/* Nagative */
.Button--intent-negative {
  background-color: #ec373c;
  color: #ffffff;
}

.Button--intent-negative:visited,
.Button--intent-negative:hover:not(:disabled),
.Button--intent-negative:link,
.Button--intent-negative:focus {
  color: #ffffff;
}

.Button--intent-negative:hover:not(:disabled) {
  /* Prevent global button and anchor styles affecting the color */
  background-color: #db151a;
  color: #ffffff;
}

/**
 * SIZES
 **/
.Button--size-extra-small {
  font-size: 0.75rem;
  height: 27px;
}

.Button--size-small {
  font-size: 0.875rem;
  line-height: 1;
  height: 28px;
}

.Button--size-medium {
  font-size: 1rem;
  height: 36px;
}

.Button--size-large {
  font-size: 1.25rem;
  line-height: 1;
  height: 40px;
}

/**
 * OTHER MODIFIERS
 **/
.Button--stretched {
  width: 100%;
}
