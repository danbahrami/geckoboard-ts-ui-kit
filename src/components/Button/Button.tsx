import React from 'react';
import noop from 'lodash/noop';
import classnames from 'classnames';
import './button.css';

export type ButtonIntent = 'primary' | 'secondary' | 'negative';
export type ButtonSize = 'extra-small' | 'small' | 'medium' | 'large';

export interface ButtonProps {
  /**
   * The content of the button. Can be a string or a JSX element.
   */
  children: React.ReactNode;

  disabled?: boolean;

  href?: string;

  intent: ButtonIntent;

  loading?: boolean;

  /**
   * How large should the button be?
   */
  size?: ButtonSize;

  /**
   * Should the button fill its parent's width
   */
  stretched?: boolean;

  target?: string;

  title?: string;

  type?: 'button' | 'reset' | 'submit';

  /**
   * Blur handler
   */
  onBlur?: (
    event: React.FocusEvent<HTMLButtonElement | HTMLAnchorElement>,
  ) => void;

  /**
   * Click handler
   */
  onClick?: (
    event: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>,
  ) => void;

  /**
   * Focus handler
   */
  onFocus?: (
    event: React.FocusEvent<HTMLButtonElement | HTMLAnchorElement>,
  ) => void;
}

export const Button = (props: ButtonProps): React.ReactElement => {
  const {
    children,
    disabled,
    href,
    intent,
    loading,
    size,
    stretched,
    target,
    title,
    type,
    onBlur,
    onClick,
    onFocus,
  } = props;

  const Tag = href ? 'a' : 'button';

  const className = classnames(
    'Button',
    `Button--intent-${intent}`,
    `Button--size-${size}`,
    { 'Button--stretched': stretched },
  );

  return (
    <Tag
      className={className}
      disabled={disabled || loading}
      href={href}
      onClick={onClick}
      onFocus={onFocus}
      onBlur={onBlur}
      target={target}
      title={title}
      type={type}
    >
      {children}
    </Tag>
  );
};

Button.defaultProps = {
  intent: 'primary',
  size: 'medium',
  type: 'button',
  onBlur: noop,
  onClick: noop,
  onFocus: noop,
};
