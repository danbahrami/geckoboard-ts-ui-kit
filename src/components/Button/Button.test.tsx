import { render, fireEvent } from '@testing-library/react';
import { Button, ButtonIntent, ButtonSize } from './Button';

describe('Components / Button', () => {
  it('renders a HTML button', () => {
    const { container } = render(<Button>Click me</Button>);

    const button = container.querySelector('button');

    expect(container).toContainElement(button);
    expect(button).toHaveAttribute('type', 'button');
    expect(button).toHaveClass('Button');
    expect(button).toHaveTextContent('Click me');
  });

  it('can render an HTML link', () => {
    const { container } = render(
      <Button href="https://www.geckoboard.com" target="_blank">
        Click me
      </Button>,
    );

    const button = container.querySelector('button');
    const link = container.querySelector('a');

    expect(container).not.toContainElement(button);
    expect(container).toContainElement(link);

    expect(link).toHaveClass(`Button`);
    expect(link).toHaveAttribute(`href`, 'https://www.geckoboard.com');
    expect(link).toHaveAttribute(`target`, '_blank');
  });

  it.each<[string, Function]>([
    ['onBlur', fireEvent.blur],
    ['onClick', fireEvent.click],
    ['onFocus', fireEvent.focus],
  ])('forwards %s events', (eventName, trigger) => {
    const spy = jest.fn();
    const props = { [eventName]: spy };

    const { container } = render(<Button {...props}>Click me</Button>);

    const button = container.querySelector('button');

    if (button) {
      trigger(button);
    } else {
      throw new Error('No button was found');
    }

    expect(spy).toHaveBeenCalled();
  });

  it('can be disabled', () => {
    const { container } = render(<Button disabled>Click me</Button>);

    const button = container.querySelector('button');

    expect(button).toBeDisabled();
  });

  it('can be a submit button', () => {
    const { container } = render(<Button type="submit">Click me</Button>);

    const button = container.querySelector('button');

    expect(button).toHaveAttribute('type', 'submit');
  });

  it('can be a reset button', () => {
    const { container } = render(<Button type="reset">Click me</Button>);

    const button = container.querySelector('button');

    expect(button).toHaveAttribute('type', 'reset');
  });

  describe.each<ButtonIntent>(['primary', 'secondary', 'negative'])(
    'when the intent is %s',
    (intent) => {
      it(`applies the --intent-${intent} CSS modifier`, () => {
        const { container } = render(<Button intent={intent}>Click me</Button>);

        const button = container.querySelector('button');

        expect(button).toHaveClass(`Button--intent-${intent}`);
      });
    },
  );

  describe.each<ButtonSize>(['extra-small', 'small', 'medium', 'large'])(
    'when the size is %s',
    (size) => {
      it(`applies the --size-${size} CSS modifier`, () => {
        const { container } = render(<Button size={size}>Click me</Button>);

        const button = container.querySelector('button');

        expect(button).toHaveClass(`Button--size-${size}`);
      });
    },
  );

  describe('when loading is true', () => {
    it('disables the button', () => {
      const { container } = render(<Button loading>Click me</Button>);

      const button = container.querySelector('button');

      expect(button).toBeDisabled();
    });
  });

  describe('snapshots', () => {
    it('matches these snapshots', () => {
      const one = render(<Button>Click me</Button>);
      expect(one.container).toMatchSnapshot();

      const two = render(<Button intent="negative">Cancel</Button>);
      expect(two.container).toMatchSnapshot();

      const three = render(
        <Button href="https://www.geckoboard.com" target="window">
          Go home
        </Button>,
      );
      expect(three.container).toMatchSnapshot();
    });
  });
});
