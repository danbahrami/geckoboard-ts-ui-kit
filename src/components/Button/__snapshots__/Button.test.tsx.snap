// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components / Button snapshots matches these snapshots 1`] = `
<div>
  <button
    class="Button Button--intent-primary Button--size-medium"
    type="button"
  >
    Click me
  </button>
</div>
`;

exports[`Components / Button snapshots matches these snapshots 2`] = `
<div>
  <button
    class="Button Button--intent-negative Button--size-medium"
    type="button"
  >
    Cancel
  </button>
</div>
`;

exports[`Components / Button snapshots matches these snapshots 3`] = `
<div>
  <a
    class="Button Button--intent-primary Button--size-medium"
    href="https://www.geckoboard.com"
    target="window"
    type="button"
  >
    Go home
  </a>
</div>
`;
